lib/Doctrine/ORM/UnitOfWork.php

```diff
@@ -1901,10 +1902,60 @@ private function doMerge($entity, array &$visited, $prevManagedCopy = null, $ass
 
         $this->cascadeMerge($entity, $managedCopy, $visited);
 
+        // Restore previous state (where the given entity was previously managed)
+        if (!$isKnown) {
```

Comment:
Can be replaced by a simple check that verifies that `$entity !== $managedCopy` or such